src/esw/science.py

```diff
@@ -0,0 +1,456 @@
+        success = self._send_msg(tx_msg)
+        return success
+
+    def _send_msg(self, tx_msg: str) -> bool:
+        """Sends the message on the UART transmit line to the STM32 chip.
+        :param tx_msg: The raw string that is to be sent to the UART transmit
+            line.
+        :returns: A boolean that is the success of sent UART transaction.
+        """
+        try:
+            tx_msg = self._add_padding(tx_msg)
+            if len(tx_msg) > self._uart_transmit_msg_len:
+                tx_msg = tx_msg[: self._uart_transmit_msg_len]
+            # rospy.loginfo(tx_msg)
+            self._uart_lock.acquire()
+            self.ser.close()
+            self.ser.open()
+            self.ser.write(bytes(tx_msg, encoding="utf-8"))
+            self._uart_lock.release()
+            return True
+        except serial.SerialException as exc:
+            if self._uart_lock.locked():
+                self._uart_lock.release()
+            rospy.logerr(f"Error in _send_msg: {exc}")
+        return False
+
+    def _servo_transmit(self, id: int, angle: float) -> bool:
+        """Sends a UART message to the STM32 chip commanding the angles of the
+        three carousel servos.
+        :param angles: Three floats in an array that represent the angles of
+            the servos. Note that this angle is what the servo interprets as
+            the angle, and it may differ from servo to servo. Also note that
+            the range of angles may vary (the safe range is between about 0
+            and 150).
+        :returns: A boolean that is the success of the transaction.
+        """
+        tx_msg = f"$SERVO,{id},{angle}"
+        success = self._send_msg(tx_msg)
+        return success
+
+    def _spectral_handler(self, tx_msg: str) -> None:
+        """Processes a UART message that contains data of the three carousel
+        spectral sensors and publishes it to the corresponding topic.
+        There are 3 spectral sensors, each having 6 channels. We read a
+        uint32_t from each channel. We know which sensor is which because we
+        know what the site is.
+        :param tx_msg: A string that was received from UART that contains data
+            of the three carousel spectral sensors.
+            - Format: <"SPECTRAL,ch_0,ch_1,....ch_5">
+        :returns: A Spectral struct that has a string has six floats that is
+        the data of the carousel spectral sensor.
+        """
+        arr = tx_msg.split(",")
+        if len(arr) < 7:
+            return
+        spectral_data = Spectral()
+        for i in range(6):
+            spectral_data.data[i] = int(arr[i + 1])
+        self._publisher_by_tag[arr[0][3:]].publish(spectral_data)
+
+
+def main():
+    rospy.init_node("science")
+    bridge = ScienceBridge()
+    rospy.Service("change_arm_laser_state", ChangeDeviceState, bridge.handle_change_arm_laser_state)
+    rospy.Service("change_auton_led_state", ChangeAutonLEDState, bridge.handle_change_auton_led_state)
```

Comment:
Sorry why do we have both this and https://github.com/umrover/mrover-ros/pull/334 ??

Reply:
yes

Reply:
334 is a temp sol